server/offenceCodeDecisions/Decision.ts

```diff
@@ -0,0 +1,191 @@
+import Code from './Code'
+
+export default class Decision {
+  private parentDecision: Decision
+
+  private childrenDecisions: Array<Decision> = new Array<Decision>()
+
+  private readonly decisionQuestion: Question
+
+  private decisionCode: Code
+
+  private decisionTitle: Title
+
+  private decisionPage: Page
+
+  private decisionUrl: string
+
+  constructor(question?: Question | string) {
+    if (question instanceof Question) {
+      this.decisionQuestion = question
+    } else if (typeof question === 'string') {
+      this.decisionQuestion = new Question(question)
+    }
+  }
+
+  id(): string {
+    if (this.getParent()) {
+      return `${this.getParent().id()}:${this.getParent().childrenDecisions.indexOf(this)}`
+    }
+    return '0'
+  }
+
+  parent(parent: Decision) {
+    this.parentDecision = parent
+  }
+
+  child(child: Decision) {
+    child.parent(this)
+    this.childrenDecisions.push(child)
+    return this
+  }
+
+  page(page: Page) {
+    this.decisionPage = page
+    return this
+  }
+
+  title(title: Title | string) {
+    if (title instanceof Title) {
+      this.decisionTitle = title
+    } else if (typeof title === 'string') {
+      this.decisionTitle = new Title(title)
+    }
+    return this
+  }
+
+  code(code: Code) {
+    this.decisionCode = code
+    return this
+  }
+
+  url(url: string) {
+    this.decisionUrl = url
+    return this
+  }
+
+  getId() {
+    return this.id()
+  }
+
+  getTitle() {
+    return this.decisionTitle
+  }
+
+  getQuestion() {
+    return this.decisionQuestion
+  }
+
+  getChildren() {
+    return this.childrenDecisions
+  }
+
+  getParent() {
+    return this.parentDecision
+  }
+
+  getUrl(): string {
+    return this.decisionUrl || this.id()
+  }
+
+  getPage(): string {
+    return this.decisionPage
+  }
+
+  getCode() {
+    return this.decisionCode
+  }
+
+  allUrls(): Array<string> {
+    const codes = [].concat(...this.getChildren().map(c => c.allUrls()))
+    if (this.getUrl()) {
+      codes.push(this.getUrl())
+    }
+    return codes.sort()
+  }
+
+  findByUrl(url: string): Decision {
+    return this.findBy(d => d.getUrl() === url)
+  }
+
+  findById(id: string): Decision {
+    return this.findBy(d => d.id() === id)
+  }
+
+  findBy(fn: (d: Decision) => boolean): Decision {
+    const matching = this.matching(fn)
+    if (matching.length !== 0) {
+      return matching[0]
+    }
+    return null
+  }
+
+  matching(fn: (d: Decision) => boolean): Array<Decision> {
+    const matches = [].concat(...this.getChildren().map(c => c.matching(fn)))
+    if (fn(this)) {
+      matches.push(this)
+    }
+    return matches
+  }
+
+  allCodes(): Array<Code> {
+    const codes = [].concat(...this.getChildren().map(c => c.allCodes()))
+    if (this.getCode()) {
+      codes.push(this.getCode())
+    }
+    return codes.sort()
+  }
+
+  invalidDecisions(): Array<Decision> {
+    return this.matching(d => d.invalid())
+  }
+
+  invalid(): boolean {
+    return (
```

Comment:
Invalid if:
We have any leaf nodes that done't have a code
We have node with children that doesn't have a title
The page or url starts with a /. This is just to ensure consistency when using them to build paths in the controllers